### ISSUE TITLE
iOS13: Sometimes driver fails to start with busy port problem

### ISSUE THREAD

**Issue by amedvedjev** [appium/appium#13452]
## The problem

After moving to iOS13 and Xcode 11 Appium Driver started getting problems on start with busy port problem.
"The port #XXXX is occupied by an other process"

Experimentally it was find out that problem happens more often (if not all time only) when we use video record during test.

## Environment
* Appium version (or git revision) that exhibits the issue: 1.16.0-beta.1
* Last Appium version that did not exhibit the issue (if applicable): unknown
* Desktop OS/version used to run Appium: macOS 10.15
* Node.js version (unless using Appium.app|exe): v12.11.1
* Npm or Yarn package manager: 6.11.3
* Mobile platform/version under test: iOS
* Real device or emulator/simulator: real iPhone

## Details
To reproduce problem faster it is enough to stop test execution in the middle of test and start again. 
in 50% we get problem that port is busy. and problem continues for 2-5min.

## Link to Appium logs
https://gist.github.com/amedvedjev/5cdd11220970995460136692559f6de7

**Comment by mykola-mokhnach** [comment]
This error is expected. Previously we were using iproxy to forward usb traffic and now the custom implementation is used, which means the Appium process itself is locking the port. There is a global (per process) class there, which manages connections. Although, it only knows about the connections it locked on its own. So, in case we don't properly quit a test (e. g. deleteSession has not been called) and kill the process then the port remains busy until timeout happens and/or the operating system itself decides to close it. There is no way to free it up from a newly created process.

Long story short - always make sure the session has been closed properly after it has been started. Or build custom ports management system...

**Comment by amedvedjev** [comment]
@mykola-mokhnach i described how to reproduce problem quicker. For us it starting right after call driver.stoprecording (fails with timeout). after this driver.quit failes and later no way to open drive for 2-3min.

If you need for logs i will add.

**Comment by KazuCocoa** [comment]
https://github.com/appium/appium/issues/13127#issuecomment-527838819 is a related issue we found before.

**Comment by amedvedjev** [comment]
@KazuCocoa random WDA port strategy did not help. still fails to start driver when it happens.

**Comment by KazuCocoa** [comment]
thanks.
I ran below test code (copy&paste them to repeat them multiply) for iOS 13 real device, but could not reproduce the error yet. (No WDA random port. Always WDA used 8100 port. useNewWDA was false when I tried.)

https://github.com/appium/ruby_lib_core/blob/362405b857fa3f36b9dfe2c745c10cd288ec577f/test/functional/ios/ios/mjpeg_server_test.rb#L21

Every time, below log appeared in create session.

```
[debug] [XCUITest] No obsolete cached processes from previous WDA sessions listening on port 8100 have been found
[DevCon Factory] Requesting connection for device 242bfcc998fd156df0bd0ad1dde8ab8e0a032114 on local port 8100, device port 8100
[debug] [DevCon Factory] Cached connections count: 0
[DevCon Factory] Port #8100 is busy
[DevCon Factory] Successfully requested the connection for 242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100
[debug] [XCUITest] Starting WebDriverAgent initialization with the synchronization key '/Users/kazu/GitHub/ruby_lib_core/tmp/9U4PT397Q4'
```

In deleting session, like below log appeared. `DevCon Factory` prefix is handling ports part.

```
[debug] [WD Proxy]   "sessionId" : "DC4BFA9A-2361-46CB-A062-038C20F9B458"
[debug] [WD Proxy] }
[DevCon Factory] Releasing connections for 242bfcc998fd156df0bd0ad1dde8ab8e0a032114 device on any port number
[DevCon Factory] Found cached connections to release: ["242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100"]
[DevCon Factory] Releasing the listener for '242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100'
[debug] [DevCon Factory] Cached connections count: 0
```

If you have a log when your test changed to work to fail (before/after reproducing the issue), it will help to observe the behaviour.

**Comment by mykola-mokhnach** [comment]
@amedvedjev Yes, I would like to see the server log where screen recording fails with timeout. Most likely the problems lies there.

**Comment by amedvedjev** [comment]
@mykola-mokhnach  will try to add.

**Comment by mykola-mokhnach** [comment]
@amedvedjev I did several improvements to the current ports cleanup logic. Please merge the PR https://github.com/appium/appium-xcuitest-driver/pull/1094 to your local source and check if it resolves the problem

**Comment by amedvedjev** [comment]
@mykola-mokhnach thanks. i can not do it now as i do not have access to remote machine from home. we try it on working days

PS USE_NEW_WDA = true - did not help

**Comment by mykola-mokhnach** [comment]
@amedvedjev Have you tried the PR? Did it work?

**Comment by vladsobolMonese** [comment]
@mykola-mokhnach 
Have tried suggested solutions but the result is exactly the same as @amedvedjev described.

**Comment by amedvedjev** [comment]
Hi. Sorry i am in vacation till 5 Nov. @vladsobolMonese is trying instead.

**Comment by mykola-mokhnach** [comment]
@vladsobolMonese Please attach the updated logs

**Comment by amedvedjev** [comment]
@mykola-mokhnach we found that we have issue: https://github.com/appium/appium/issues/13454 and continue investigating to see if we get something more ...

**Comment by vladsobolMonese** [comment]
@mykola-mokhnach 
Noticed in logs that we are getting this error
`org.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. Original error: The screen capture process 'ffmpeg' died unexpectedly`

https://gist.github.com/vladsobolMonese/5be06df7683e1bc86f7112cd11d714de

After this tests keep running and at some point get this failure
```
Element [[io.appium.java_client.ios.IOSDriver, Capabilities: {app=/Users/qa/Documents/test/app/SpecificApp.ipa, automationName=XCuiTest, browserName=, bundleId=com.monese.dev, clearSystemFiles=true, databaseEnabled=false, deviceName=iPhone QA4, fullReset=true, javascriptEnabled=true, locationContextEnabled=false, mjpegServerPort=6204, nativeWebTap=true, networkConnectionEnabled=false, newCommandTimeout=180, noReset=false, platform=MAC, platformName=iOS, platformVersion=13.0, processArguments={args=[-featureFlag.tp_joint_account_cards, YES, -featureFlag.CRD: Virtual Card, YES, -featureFlag.RW: New Referral Program, YES], env={key=value}}, sendKeyStrategy=setValue, simpleIsVisibleCheck=true, takesScreenshot=true, udid=a282119bd8db78f1d8ea2052b0f93b8e45c860d6, useJSONSource=true, wdaLaunchTimeout=120000, wdaLocalPort=8002, wdaStartupRetries=4, webStorageEnabled=false, xcodeOrgId=22WULF6P39, xcodeSigningId=iPhone Developer}] -> id: switchUserButton]
Connection refused (Connection refused)
Build info: version: '4.0.0-alpha-1', revision: 'd1d3728cae', time: '2019-04-24T13:42:21'
System info: host: 'qa's Mac', ip: 'fe80:0:0:0:72:a59f:6135:b194%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.15', java.version: '1.8.0_201'
Driver info: driver.version: IOSDriver
```

And in the end of this test we try to stopVideoRecording we get
`org.openqa.selenium.WebDriverException: Connection refused (Connection refused)`

**Comment by amedvedjev** [comment]
@mykola-mokhnach issues almost fix with latest beta! there is still a bit problem happens sometimes when we try to get pageSource at the end of test. Same error:  "socket hang up" BUT driver quit and next driver start = zero problem. i willget appium logs for this case later.

**Comment by amedvedjev** [comment]
looks like issue gone with latest build

**Comment by zcmgyu** [comment]
My solution is restarting the iPhone device. Or delete target `DerivedData` pointed out in logs.

**Example:**
```
rm -rf /Users/username/Library/Developer/Xcode/DerivedData/WebDriverAgent-ckoduscgxudzpehatfkuzhqvbpqx
```
